ccips/ccip-009/ccip-009-citycoins-vrf-v2.md

```diff
@@ -0,0 +1,214 @@
+# CCIP-009
+
+## Preamble
+
+| CCIP Number   | 009                                   |
+| ------------- | ------------------------------------- |
+| Title         | CityCoins VRF v2                      |
+| Author(s)     | Jason Schrader jason@joinfreehold.com |
+| Consideration | Technical                             |
+| Type          | Standard                              |
+| Status        | Activation-In-Progress                |
+| Created       | 2022-03-25                            |
+| License       | BSD-2-Clause                          |
+| Replaces      | CCIP-006                              |
+
+## Introduction
+
+The Stacks blockchain is a Layer 1 blockchain connected to Bitcoin, in which miners spend Bitcoin to bid for and win a fixed amount of Stacks tokens. Stackers have the option to lock up Stacks tokens for a specified amount of time, and in turn, receive a portion of the Bitcoin spent by miners proportionate to the amount Stacked[^1].
+
+Taking this concept a level further, a new token (a "CityCoin") is created on the Stacks blockchain, following the SIP-010 fungible token standard[^2], such that CityCoins can be mined and Stacked per the methods stated above except a portion of the miner's bid would be redirected to the city's wallet overseen by a trusted third party custodian.
+
+CityCoins leverage similar properties from the Proof of Transfer (PoX) consensus mechanism[^3] of the Stacks blockchain, programmed through a smart contract in Clarity.
+
+This proposal describes the verifiable random function (VRF) used to calculate the block winner for a CityCoin.
+
+## Specification
+
+The VRF contract is used to determine the winner of a mined block.
+
+This version introduces two functions to replace `get-random-uint-at-block` in CCIP-006:
+
+- `get-rnd`: a read-only function that returns the random integer
+- `get-save-rnd`: a public function that saves the random integer to a map, then returns it
+
+Through the `get-rnd` function, the contract:
+
+- checks the `BlockRnd` map to see if the random integer has been saved
+  - if yes, returns the saved random integer
+  - if no, it then:
+    - fetches the on-chain VRF proof for the given block height
+    - converts the lower 16 bytes into a uint
+    - returns the random integer
+
+Through the `get-save-rnd` function, the contract:
+
+- checks the `BlockRnd` map to see if the random integer has been saved
+  - if yes, returns the saved random integer
+  - if no, it then:
+    - fetches the on-chain VRF proof for the given block height
+    - converts the lower 16 bytes into a uint
+    - saves the random integer to the `BlockRnd` map
+    - returns the random integer
+
+The public function requires a private key and a Stacks transaction fee in order to send the transaction, and would ideally be used by any smart contract using this implementation to obtain a random value at a given block height.
```

Comment:
This sentence could be misunderstood, that you have to provide your private key in order to get value from public function.
I would remove it.

Reply:
Removed in https://github.com/citycoins/governance/pull/5/commits/f4f44b8e6e3cc5cb7ef68d215c29c2cf1676f06f